rwengine/src/objects/VehicleObject.cpp

```diff
@@ -307,11 +323,78 @@ void VehicleObject::tickPhysics(float dt) {
-        float engineForce = info->handling.acceleration * throttle * velFac;
+        float velocity = collision->getBulletBody()->getLinearVelocity().length();
+        float velocityForward = physVehicle->getCurrentSpeedKmHour() / 3.6f;
+        float velocityMax = velFac / 9.f;
+        float steerValue = 0.f;
+        float steerLimit = glm::radians(info->handling.steeringLock);
+        // The engine force is calculated from the acceleration and max velocity
+        // of the vehicle, with a specific coefficient to make it adapted to
+        // Bullet physics and avoid reaching top speed too fast.
+        float engineForce = info->handling.acceleration * throttle * velFac / 1.2f;
         float brakeF = getBraking();
+        // Mass coefficient, that quantifies how heavy a vehicle is and excludes
+        // light vehicles.
+        float kM = (std::max(1500.f, info->handling.mass) - 1500.f) / 1500.f;
+        unsigned int count = 0;
+
+        // Get the global vehicle steering value (for front wheels only).
+        for (int w = 0; w < physVehicle->getNumWheels(); ++w) {
+            btWheelInfo& wi = physVehicle->getWheelInfo(w);
+
+            if (wi.m_bIsFrontWheel) {
+                steerValue += physVehicle->getSteeringValue(w);
+                count++;
+            }
+        }
+
+        steerValue /= count;
+
+        // Increase the engine force based on the mass by up to 4 times.
+        // Heavy vehicles need extra engine force to be reactive enough.
+        engineForce *= std::min(1.f + kM, 4.f);
+
+        // Give vehicles a boost when starting (forward or backward) to reduce
+        // general sluggishness. The engine force is multiplied by 1.5 before
+        // reaching velocityMax / 4 and then reduces down to nominal force as
+        // velocity reaches velocityMax / 2.
+        if (velocity < velocityMax / 4.f)
+            engineForce *= 1.5f;
+        else if (velocity < velocityMax / 2.f)
+            engineForce *= 1.f + 0.5f * (2.f - velocity / (velocityMax / 4.f));
+
+        // Reduce the engine force when steering, by a factor of up to 1.25 for
+        // the maximum steering angle, linearly back to nominal value when no
+        // steering is applied.
+        if (std::abs(steerValue) > steerLimit / 8.f && velocity > velocityMax / 3.f)
+            engineForce /= 1.f + std::abs(steerValue) / (4.f * steerLimit);
+
+        if (velocity > velocityMax) {
+            btVector3 v = collision->getBulletBody()->getLinearVelocity().normalized();
+
+            velocity = velocityMax;
+            v *= velocity;
+
+            collision->getBulletBody()->setLinearVelocity(v);
+        }
 
         if (handbrake) {
             brakeF = 5.f;
+        } else if (throttle < 0.f && velocityForward > velocityMax / 8.f) {
+            engineForce = 0.f;
+            brakeF = 2.f * std::min(1.f + kM, 4.f);
+        }
+
+        if (isStopped() && throttle == 0.f) {
```

Comment:
I guess you should use something like : `throttle <= 0.1f`

Reply:
I don't think that would be right. Throttle is set to an integer value (either -1.f, 0.f or 1.f) so there is about no need for setting a threshold. In addition, what we're really interested in here is the velocity, which is checked by isStopped. The throttle condition is really just a way to allow the car to start again (it can hardly happen if the x coordinate of the velocity vector remains 0, since it makes it hard for the two other coordinate to compensate until reaching the point where isStopped() == false). When there is no driver interaction with the vehicle, throttle is effectively equal to zero.

Reply:
I was thinking more about problem: `comparing floating point with == or != is unsafe`.

Reply:
Integer value, no controller support?

Reply:
Okay, fair enough, I'll change that.